### PR TITLE
Make eval runs non-blocking; `spice.eval.{results, runs}` tables.

### PR DESCRIPTION
**Pull request by Jeadie**
# 🗣 Description
 - Update `/v1/evals/:name` to handle eval asynchronous. This is done via a new `EvalWorker` struct that reads new evals from a channel.
 - Two new tables:
   - `spice.eval.runs`: Tracks the details of a specific eval run (e.g. call to `POST /v1/evals/:name`).
   - `spice.eval.results`: Tracks the specific results within an eval run. This is the output of each input/output pair of an eval dataset, and their score. 
 - `SinkConnector` does not yet supports primary keys, limiting `spice.eval.runs` (it will append all state changes, and not update the row). 

## 🔨 Related Issues
 - #3320

### CODE REVIEW

File: crates/runtime/src/init/mod.rs [hunk 1]
```diff
@@ -19,7 +19,8 @@ limitations under the License.
 pub(crate) mod catalog;
 pub(crate) mod dataset;
 pub(crate) mod embedding;
-pub(crate) mod eval_scorer;
+// #[cfg(feature = "models")]
```
Comment (phillipleblanc) [inline crates/runtime/src/init/mod.rs]: ?

File: crates/runtime/src/internal_table.rs [hunk 2]
```diff
@@ -76,7 +77,17 @@ async fn get_local_table_provider(
...
     dataset.mode = Mode::ReadWrite;
 
-    let data_connector = Arc::new(SinkConnector::new(Arc::clone(schema))) as Arc<dyn DataConnector>;
+    let mut sink = SinkConnector::new(Arc::clone(schema));
+    if let Some(pk) = primary_key {
+        println!(
+            "We're adding {:?} to {:?}",
+            pk.clone(),
+            name.to_quoted_string()
+        );
```
Comment (phillipleblanc) [inline crates/runtime/src/internal_table.rs]:
```suggestion
```

File: crates/runtime/src/internal_table.rs [hunk 3]
```diff
@@ -87,22 +98,30 @@ async fn get_local_table_provider(
...
-
+    let source_table_provider =
+        get_local_table_provider(&name, &schema, primary_key.clone()).await?;
+    let federated_table = Arc::new(FederatedTable::new(Arc::clone(&source_table_provider)));
+    println!(
+        "create_internal_accelerated_table: name: {:?} == {:?} - {:?}",
+        name.to_quoted_string(),
+        primary_key.clone(),
+        Arc::clone(&source_table_provider).constraints(),
+    );
```
Comment (phillipleblanc) [inline crates/runtime/src/internal_table.rs]:
```suggestion
```

File: crates/runtime/src/model/eval/dataset.rs [hunk 4]
```diff
@@ -34,51 +37,10 @@ use async_openai::{
...
-use llms::chat::Chat;
 use serde::{Deserialize, Serialize};
 use serde_json::Value;
-use snafu::{ResultExt, Snafu};
+use snafu::ResultExt;
 use spicepod::component::eval::Eval;
-use std::{collections::HashMap, sync::Arc};
-
-#[derive(Debug, Snafu)]
-pub enum Error {
```
Comment (Jeadie) [inline crates/runtime/src/model/eval/dataset.rs:45]: These errors have just been moved to `crates/runtime/src/model/eval/mod.rs`.

File: crates/runtime/src/model/eval/mod.rs [hunk 5]
```diff
@@ -0,0 +1,294 @@
...
+                score[i],
+            )?;
+        }
+    }
+
+    write_result_to_table(Arc::clone(&df), run_id, &mut bldr).await
+}
+
+/// Return format of [`DatasetOutput`] determined by `output_format`. `output_format` can be empty, is only used for its enum type.
+#[allow(clippy::borrowed_box)]
```
Comment (phillipleblanc) [inline crates/runtime/src/model/eval/mod.rs]:
```suggestion
```

File: crates/runtime/src/model/eval/runs.rs [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+        Arc::new(StringArray::from(vec![status.to_string()])),
+    );
+    if let Some(err) = err_msg {
+        updates.insert("error_message", Arc::new(StringArray::from(vec![err])));
+    };
+    update_eval_run(df, id, updates).await
+}
+
+/// Writes a new row to `spice.evals.runs` table and returns primary key.
+pub async fn start_eval_run(
```
Comment (phillipleblanc) [inline crates/runtime/src/model/eval/runs.rs]:
```suggestion
pub async fn start_tracing_eval_run(
```

I think this is clearer on what this function actually does

**Review by phillipleblanc — DISMISSED** [review]

**Review by phillipleblanc — APPROVED** [review]

**Pull request merged**